No root struct terminator

jaekwon opened the pull request:
Also, fixes README.
Does not encode "default values", and in nillable lists considers them nil.
This is a checkpoint before diving into proto3 compatibility.

jaekwon committed:
Only include prefix/type-wrapper in Marshal* or for interface values (not otherwise)

jaekwon committed:
opts -> fopts

jaekwon committed:
Remove struct terminator; "Void"->DefaultValue; README fixes

liamsi commented on binary-decode.go:
```diff
@@ -599,6 +599,13 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
 
+			// We're done if isRoot & we've consumed all the bytes.
+			// NOTE: any unseen fields are zero, consistent
+			// with the continue statement above.
```
Do you mean the continue statement below? Or another function above (which?)?

jaekwon replied:
Yeah I removed this in a later PR.

liamsi commented on binary-decode.go:
```diff
@@ -611,7 +618,8 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
 			if field.BinFieldNum < fieldNum {
-				// Set nil field value.
+				// Set zero field value.
+				// NOTE: See note on isRoot below in this function.
```
Are you referring to the comment above? Or the one on `if !isRoot {`?

liamsi commented on binary-decode.go:
```diff
@@ -637,24 +645,27 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
+		// If not root object, read "StructTerm".
+		if !isRoot {
+			// NOTE: In the future, we'll need to break out of a loop
+			// when encoutering an StructTerm typ3 byte.
```
s/encoutering/encountering

liamsi commented on codec.go:
```diff
@@ -79,12 +79,13 @@ type InterfaceOptions struct {
-	ConcreteOptions              // Registration options.
+	Registered       bool // Registered with RegisterConcrete().
+	PointerPreferred bool // Deserialize to pointer type if possible.
+	// NilPreferred     bool        // Deserialize to nil for empty structs if PointerPreferred.
```
remove if not used

liamsi commented on reflect.go:
```diff
@@ -72,14 +72,22 @@ func derefPointers(rv reflect.Value) (drv reflect.Value, isPtr bool, isNilPtr bo
+		case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+			return rv, rv.Int() == 0
+		case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+			return rv, rv.Uint() == 0
```
We should keep an eye on this for amino_rs too /cc @zmanian This might have some minor implications on the kms code for instance.

jaekwon replied:
Yeah.  This should be (in later PRs at least) compatible with what Proto3 does.

liamsi reviewed (commented):
LGTM with some nitpicks: one/two comments are slightly confusing  (what do they refer to); typo

jaekwon committed:
WIP proto3 compatibility (#163)

This PR makes Amino:binary compatible with Proto3 (but not Proto2). It doesn't implement all of the features of Proto3, but everything should be parseable in Proto3.

Notable differences are, support of Interfaces as byte-length prefixed blobs. Support for nested arrays is also interesting.

This makes the encoder slower, but the purpose is to start with Proto3 compatibility and work toward Amin2.

Many thanks to @ebuchman and @mdyring for exploring this.

jaekwon committed:
Merge branch 'develop' into jae/norootterm
